contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,594 @@
+then use that as the response.
+
+```
+POST [arbitrary_url]
+{
+    "kind": "AdmissionReview",
+    "apiVersion": "admission.k8s.io/v1",
+    "spec": {
+        "resource": "pods",
+        "subresource": "",
+        "operation": "Create",
+        "object": {...},
+        "oldObject": nil,
+        "userInfo": {
+            "name": "user1",
+            "uid": "cn=user,dn=ou",
+            "groups": ["g=1"]
+        }
+    },
+    "status": {
+        "status": "Failure",
+        "message": "...",
+        "reason": "Forbidden",
+        "code": 403,
+        ...
+    }
+}
+
+400 Bad Request
+{
+    "kind": "AdmissionReview",
+    "apiVersion": "admission.k8s.io/v1",
+    "spec": {
+        ...
+    },
+    "status": {
+        "status": "Failure",
+        "message": "...",
+        "reason": "Forbidden",
+        "code": 403,
+        ...
+    }
+}
+```
+
+Clients may return 204 as an optimization. `oldObject` may be sent if this is an update operation. This
+API explicitly does not allow mutations, but leaves open the possibility for that to be added in the future.
+
+Each webhook must be non-mutating and will be invoked in parallel. The first failure will terminate processing,
+and the caller may choose to retry calls and so submission must be idempotent.
+
+Because admission is performance critical, the following considerations are taken:
+
+1. The caller may pass the received body bytes AS-IS to the admission controller, so defaulting may not be performed.
+2. Protobuf will be the expected serialization format - JSON is shown above for readability, but the content-type wrapper will be used to encode bytes.
+3. The admission object will be serialized once and sent to all callers
+4. To minimize fan-out variance, future implementations may set strict timeouts and dispatch multiple requests.
+
+Admission is a high security operation, so end-to-end TLS encryption is expected and the remote endpoint
+should be authorized via strong signing, mutual-auth, or high security.
+
+
+##### Bypassing external admission hooks
+
+There may be scenarios where an external admission hook blocks a system critical loop in a non-obvious way -
+by preventing node updates that prevents a new admission pod from being created, for instance. One option
+is to allow administrators to request fail open on specific calls, or to require that certain special resource
+paths (initializers dynamic config path) are always fail open. Alternatively, it may be desirable for
+administrative users to bypass admission completely.
+
+Some options:
+
+1. Some namespaces are opted out of external admission (kube-system)
+2. Certain service accounts can bypass external admission checks
+
+
+##### Upgrade of a cluster with external admission
+
+The current order of cluster upgrade is `apiserver` -> `controller` -> `nodes`.  External admission controllers
+would typically need to be upgraded first in order to ensure new semantic changes in objects are not ignored.
```

Comment:
Not sure whether you mean real controllers here only or also the webhooks. The later have to be able to decode the **not** negotiated payload, which enforces certain upgrade orders.

Not sure you mentioned this explicitly anywhere: what's the payload version for webhooks? The prefered version? Configurable?